Pokemon

ijassiem opened the pull request.

ijassiem committed:
Added mod,divide,multiply

ijassiem committed:
Added functions and unit tests DGDT-14

ijassiem committed:
Create a pokemon class for instantiating pokemon characters.

This change consists of a base Pokemon class and a derived ElectricPokemon class.

These classes create pokemon characters. Methods are included for allowing the
pokemon character to speak, learn a new move, and print all its properties.
Unit tests for the classses are included.

Resolves: "DGDT-19"

james-smith-za commented on pokemon/pokemon.py:
```diff
@@ -0,0 +1,79 @@
+
+
+class Pokemon(object):
+    """
+    This class Pokemon represents a pokemon character with several attributes.
+
+     :param name: Name of pokemon
+     :type name: str
+     :param nickname: Nickname of pokemon
+     :type nickname: str
+     :param moves: list of 4 moves
+     :type moves: list
+    """
+
+    def __init__(self, name, nickname):
+        """Initialise Pokemon class."""
+        self.name = name
+        self.nickname = nickname
+        self.moves = ["jump", "strike", "dash", "block"]
```
Couple of points:

1. This docstring format is valid, and perfectly fine if you're working on your own. We have an informal convention though, to use numpy-style docstrings, all of our other modules do it. See here for the format: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html
2. I prefer not to tell the type of the parameters in the docstring. Recent versions of Python support type hints, as I've indicated in the suggestion below. I find that this is better practice for two reasons:
  a. `mypy` can work with it and help you out as you code, and 
  b. when you're maintaining the code later, and you need to modify something, it's much easier to update the type hint and keep it correct when it's right there in your code, rather than in the docstring.
3.  "Parameters" strictly means the things between the brackets `(name, nickname)`. You have another one, `moves`, which gets initialised but it isn't one of the parameters that you pass to the `__init__` method. You can still document this, but you would put it under the heading of `Attributes`. I know how to do that with Numpy style, I'm not really sure about the convention that you've used here.

```suggestion
    """
    This class Pokemon represents a pokemon character with several attributes.

     :param name: Name of pokemon
     :param nickname: Nickname of pokemon
    """

    def __init__(self, name: str, nickname: str):
        """Initialise Pokemon class."""
        self.name = name
        self.nickname = nickname
        self.moves = ["jump", "strike", "dash", "block"]
```

james-smith-za commented on pokemon/pokemon.py:
```diff
@@ -0,0 +1,79 @@
+
+    def learn_move(self, new_move):
+        """Pokemon learn a new move."""
+        self.moves.append(new_move)
+        self.moves.pop(0)
```
Today I learned that lists have `pop()` methods. I wasn't aware of that.

Just be careful how you use this. What is happening here is that you are appending a new move, and then popping the same new move right off the list!

If you want to push one of the old ones out, consider making the list of moves a `deque` rather (you need to `from collections import deque`) and then you can `popleft()`.

james-smith-za replied:
Ooooh! Never mind, disregard this comment. I didn't notice the `0` initially, that specifically pops element number `0`/

My bad, the way you have it will work perfectly.

james-smith-za commented on pokemon/pokemon.py:
```diff
@@ -0,0 +1,79 @@
+p.speak()
+p.print_details()
+print("")
+p.learn_move("glide")
+p.print_details()
```
So you'll have noticed here that despite learning "glide", your Pikachu will still have the same four moves as before.

james-smith-za commented on testing/test_pokemon.py:
```diff
@@ -0,0 +1,63 @@
+"""Test module."""
+
+import pokemon.pokemon as p
+import unittest
+import pytest
```
One wouldn't normally use both of these as far as I am aware? Just one or the other. From the way you've structured things below it looks as though pytest is the one you wane to keep.

ijassiem replied:
The main reason I included unittest was to use assertAlmostEqual. I need to find something similar in pytest.

james-smith-za commented on testing/test_pokemon.py:
```diff
@@ -0,0 +1,63 @@
+        my_pokemon = p.ElectricPokemon("slowpoke", "poke", "psychic")
+        self.assertIsInstance(my_pokemon, p.ElectricPokemon)
+
+    def test_learn_move(self):
+        """Test a new move is learnt."""
+        my_pokemon = p.Pokemon("charmander", "char")
+        my_pokemon.learn_move("slice")
+        assert my_pokemon.moves[3] == "slice"
```
I'd be interested to see what the outcome of this test is.

james-smith-za commented:
Aha, understood. `np.isclose()` might be something you could use, but it involves using numpy.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Monday, June 7th, 2021 at 11:11, Ismail ***@***.***> wrote:

> @ijassiem commented on this pull request.
> 

> In testing/test_pokemon.py:
> 

> > +import unittest
> +import pytest
> 

> The main reason I included unittest was to use assertAlmostEqual. I need to find something similar in pytest.
> 

> —
> 

> You are receiving this because your review was requested.
> 

> Reply to this email directly, view it on GitHub, or unsubscribe.

ijassiem committed:
Implement changes relating to comments in pull request

These changes pertains to the comments indicated in the pull request: "Pokemon #3".

Added numpy style docstrings to pokemon.py.

Contributes to: DGDT-19
